_posts/2020-02-23-pyflink-udf-support-flink.md

```diff
@@ -0,0 +1,280 @@
+
+Finally, you can see the execution result on the command line:
+
+`$ cat /tmp/output
+ 3`
+
+
+### Python UDF dependency management
+
+In many cases, you would like to import third-party dependencies in the Python UDF. The example below provides detailed guidance on how to manage such dependencies.
+
+Suppose you want to use the `mpmath` to perform the sum of the example above. The Python UDF may look like:
+
+```python
+@udf(input_types=[DataTypes.BIGINT(), DataTypes.BIGINT()], result_type=DataTypes.BIGINT())
+def add(i, j):
+    from mpmath import fadd # add third-party dependency
+    return int(fadd(1, 2))
+```
+
+To make it available on the worker node that does not contain the dependency, you can specify the dependencies with the following API:
+
+```python
+# echo mpmath==1.1.0 > requirements.txt
+# pip download -d cached_dir -r requirements.txt --no-binary :all:
+t_env.set_python_requirements("/path/of/requirements.txt", "/path/of/cached_dir")
+```
+
+A `requirements.txt` file that defines the third-party dependencies is used. If the dependencies cannot be accessed in the cluster, then you can specify a directory containing the installation packages of these dependencies by using the parameter "`requirements_cached_dir`", as illustrated in the example above. The dependencies will be uploaded to the cluster and installed offline. 
+
+Below, we showcase a complete example of using dependency management for Python UDFs in Flink 1.10.
+
+```python
+from pyflink.datastream import StreamExecutionEnvironment
+from pyflink.table import StreamTableEnvironment, DataTypes
+from pyflink.table.descriptors import Schema, OldCsv, FileSystem
+from pyflink.table.udf import udf
+
+env = StreamExecutionEnvironment.get_execution_environment()
+env.set_parallelism(1)
+t_env = StreamTableEnvironment.create(env)
+
+@udf(input_types=[DataTypes.BIGINT(), DataTypes.BIGINT()], result_type=DataTypes.BIGINT())
+def add(i, j):
+    from mpmath import fadd
+    return int(fadd(1, 2))
+
+t_env.set_python_requirements("/tmp/requirements.txt", "/tmp/cached_dir")
+t_env.register_function("add", add)
+
+t_env.connect(FileSystem().path('/tmp/input')) \
+    .with_format(OldCsv()
+                 .field('a', DataTypes.BIGINT())
+                 .field('b', DataTypes.BIGINT())) \
+    .with_schema(Schema()
+                 .field('a', DataTypes.BIGINT())
+                 .field('b', DataTypes.BIGINT())) \
+    .create_temporary_table('mySource')
+
+t_env.connect(FileSystem().path('/tmp/output')) \
+    .with_format(OldCsv()
+                 .field('sum', DataTypes.BIGINT())) \
+    .with_schema(Schema()
+                 .field('sum', DataTypes.BIGINT())) \
+    .create_temporary_table('mySink')
+
+t_env.from_path('mySource')\
+    .select("add(a, b)") \
+    .insert_into('mySink')
+
+t_env.execute("tutorial_job")
+```
+
+### Submit the job
```

Comment:
we had the same heading before. This is a bit confusing

Reply:
May be `Submit a job with dependency` ?